File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/ModalPlacementSelector.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.property;

import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonList;
import com.urbanairship.json.JsonMap;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class ModalPlacementSelector {
    @NonNull
    private final ModalPlacement placement;
    @Nullable
    private final WindowSize windowSize;
    @Nullable
    private final Orientation orientation;

    public ModalPlacementSelector(
        @NonNull ModalPlacement placement,
        @Nullable WindowSize windowSize,
        @Nullable Orientation orientation) {
        this.placement = placement;
        this.windowSize = windowSize;
        this.orientation = orientation;
    }

    @NonNull
    public static ModalPlacementSelector fromJson(@NonNull JsonMap json) throws JsonException {
        JsonMap placementJson = json.opt("placement").optMap();
        String windowSizeString = json.opt("window_size").optString();
        String orientationString = json.opt("orientation").optString();

        ModalPlacement placement = ModalPlacement.fromJson(placementJson);
        WindowSize windowSize = windowSizeString.isEmpty() ? null : WindowSize.from(windowSizeString);
        Orientation orientation = orientationString.isEmpty() ? null : Orientation.from(orientationString);

        return new ModalPlacementSelector(placement, windowSize, orientation);
    }

    @NonNull
    public static List<ModalPlacementSelector> fromJsonList(@NonNull JsonList json) throws JsonException {
        List<ModalPlacementSelector> selectors = new ArrayList<>(json.size());
        for (int i = 0; i < json.size(); i++) {
            JsonMap selectorJson = json.get(i).optMap();
            ModalPlacementSelector selector = ModalPlacementSelector.fromJson(selectorJson);
            selectors.add(selector);
        }
        return selectors;
    }

    @NonNull
    public ModalPlacement getPlacement() {
        return placement;
    }

    @Nullable
    public WindowSize getWindowSize() {
        return windowSize;
    }

    @Nullable
    public Orientation getOrientation() {
        return orientation;
    }
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/ModalPlacementSelector.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.property

import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonList
import com.urbanairship.json.JsonValue

public class ModalPlacementSelector public constructor(
    @JvmField public val placement: ModalPlacement,
    @JvmField public val windowSize: WindowSize?,
    @JvmField public val orientation: Orientation?
) {

    public companion object {
        private const val KEY_PLACEMENT = "placement"
        private const val KEY_WINDOW_SIZE = "window_size"
        private const val KEY_ORIENTATION = "orientation"

        @Throws(JsonException::class)
        public fun fromJson(json: JsonValue): ModalPlacementSelector {
            val content = json.requireMap()

            return ModalPlacementSelector(
                placement = ModalPlacement.fromJson(content.require(KEY_PLACEMENT)),
                windowSize = content[KEY_WINDOW_SIZE]?.let(WindowSize::from),
                orientation = content[KEY_ORIENTATION]?.let(Orientation::from)
            )
        }

        @JvmStatic
        @Throws(JsonException::class)
        public fun fromJsonList(json: JsonList): List<ModalPlacementSelector> {
            return json.map(::fromJson)
        }
    }
}
